kernel/src/table_changes/scan.rs

```diff
@@ -188,6 +187,112 @@ impl TableChangesScan {
+    /// to read and process all the data for the query. Each [`ScanResult`] in the resultant iterator
+    /// encapsulates the raw data and an optional boolean vector built from the deletion vector if it
+    /// was present. See the documentation for [`ScanResult`] for more details.
+    pub fn execute(
+        &self,
+        engine: Arc<dyn Engine>,
+    ) -> DeltaResult<impl Iterator<Item = DeltaResult<ScanResult>>> {
+        let scan_data = self.scan_data(engine.clone())?;
+        let scan_files = scan_data_to_scan_file(scan_data);
+
+        let global_scan_state = self.global_scan_state();
+        let table_root = self.table_changes.table_root().clone();
+        let all_fields = self.all_fields.clone();
+        let predicate = self.predicate.clone();
+        let dv_engine_ref = engine.clone();
+
+        let result = scan_files
+            .map(move |scan_file| {
+                resolve_scan_file_dv(dv_engine_ref.as_ref(), &table_root, scan_file?)
+            }) // Iterator-Result-Iterator
+            .flatten_ok() // Iterator-Result
+            .map(move |resolved_scan_file| -> DeltaResult<_> {
+                read_scan_file(
+                    engine.as_ref(),
+                    resolved_scan_file?,
+                    &global_scan_state,
+                    &all_fields,
+                    predicate.clone(),
+                )
+            }) // Iterator-Result-Iterator-Result
+            .flatten_ok() // Iterator-Result-Result
+            .map(|x| x?); // Iterator-Result
+
+        Ok(result)
+    }
+}
+
+/// Reads the data at the `resolved_scan_file` and transforms the data from physical to logical.
+/// The result is a fallible iterator of [`ScanResult`] containing the logical data.
+fn read_scan_file(
+    engine: &dyn Engine,
+    resolved_scan_file: ResolvedCdfScanFile,
+    global_state: &GlobalScanState,
+    all_fields: &[ColumnType],
+    predicate: Option<ExpressionRef>,
+) -> DeltaResult<impl Iterator<Item = DeltaResult<ScanResult>>> {
+    let ResolvedCdfScanFile {
+        scan_file,
+        mut selection_vector,
+    } = resolved_scan_file;
+
+    let physical_to_logical_expr =
+        physical_to_logical_expr(&scan_file, global_state.logical_schema.as_ref(), all_fields)?;
+    let read_schema = scan_file_read_schema(&scan_file, global_state.read_schema.as_ref());
+    let phys_to_logical_eval = engine.get_expression_handler().get_evaluator(
+        read_schema.clone(),
+        physical_to_logical_expr,
+        global_state.logical_schema.clone().into(),
+    );
+
+    let table_root = Url::parse(&global_state.table_root)?;
+    let location = table_root.join(&scan_file.path)?;
+    let file = FileMeta {
+        last_modified: 0,
+        size: 0,
+        location,
+    };
+    let read_result_iter =
+        engine
+            .get_parquet_handler()
+            .read_parquet_files(&[file], read_schema, predicate)?;
```

Comment:
(don't have to solve now) I wonder if we should introduce this higher in the iterator pipeline we've built up here?

Oversimplification, but instead of something like:
```rust
scan_files
    .map(|scan_file| read_parquet_files([scan_file])) // parquet batches
    .map(|batch| do_stuff) // return final batches
```
could push it up to be
```rust
read_parquet_files(scan_files)  // parquet batches
    .map(|batch| do_stuff) // return final batches
```

Reply:
unsure how much state you'd need to track and associate with each batch (if it makes things too complicated)

Reply:
Hmm we'd have to somehow associate the `scan_file` and `selection_vector` to the `batch` because it's important for the P2L transformation.  So the type of the proposed `read_parquet_files` would have to be something like `Iterator<(EngineData, ResolvedCdfScanFile)>`. 
and `do_stuff: (EngineData, ResolvedCdfScanFile) -> ScanResult`

Reply:
oh i posted another comment about this, but maybe just make an issue to invest a bit more time into this